product.py

```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
         self.mount_state()
         result = True
         try:
+            # Don't propose to upgrade a BIOS installation with a UEFI installer and conversely
+            tool = PartitionTool(self.primary_disk)
+            existing_boot_mode = constants.TARGET_BOOT_MODE_LEGACY
+            boot_partnum = tool.partitionNumber(self.boot_device)
+            boot_part = tool.getPartition(boot_partnum)
+            if 'id' in boot_part and boot_part['id'] == GPTPartitionTool.ID_EFI_BOOT:
+                existing_boot_mode = constants.TARGET_BOOT_MODE_UEFI
+
+            if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI and not constants.UEFI_INSTALLER) or \
+                (existing_boot_mode == constants.TARGET_BOOT_MODE_LEGACY and constants.UEFI_INSTALLER):
```

Comment:
a bit heavy to read, what about something like this (closer to XOR)
```
if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI) != constants.UEFI_INSTALLER:
```

Will also allow to get a visually clear `if` statement without a double indent (never seen that in the wild, pretty sure the code checkers would flag it)

Reply:
Done